### ISSUE TITLE
Avoid runtime code compilation for performance improvement

### ISSUE THREAD

**Issue by xerial** [wvlet/airframe#25]
When binding a trait:
```scala
    .bind[StatementHandler].toEagerSingletonOf[DefaultStatementHandler]
```

If the trait has no Session, we need to dynamically embed a Session object to it:
```
2016-09-08 10:10:41-0700 trace [wvlet.airframe.SessionImpl] Search bindings of StatementHandler  - (SessionImpl.scala:105)
2016-09-08 10:10:41-0700 trace [wvlet.airframe.SessionImpl] Found a singleton for StatementHandler: DefaultStatementHandler  - (SessionImpl.scala:115)
2016-09-08 10:10:41-0700 trace [wvlet.airframe.SessionImpl] Search bindings of DefaultStatementHandler  - (SessionImpl.scala:105)
2016-09-08 10:10:41-0700 trace [wvlet.airframe.SessionImpl] No binding is found for DefaultStatementHandler  - (SessionImpl.scala:141)
2016-09-08 10:10:41-0700 trace [wvlet.airframe.SessionImpl] buildInstance DefaultStatementHandler, stack:List(DefaultStatementHandler, StatementHandler)  - (SessionImpl.scala:148)
2016-09-08 10:10:41-0700 trace [wvlet.airframe.SessionImpl]
Compiling a code for embedding Session to DefaultStatementHandler:
new (wvlet.airframe.Session => Any) {
  def apply(session:wvlet.airframe.Session) = {
    new com.treasuredata.prestobase.proxy.StatementHandler.DefaultStatementHandler {
      protected def __current_session = session
    }
  }
}    - (SessionImpl.scala:184)
2016-09-08 10:10:42-0700 trace [wvlet.airframe.Session] Checking a session for class __wrapper$1$7ff11152d6594d33b9c958a6a4b57872.__wrapper$1$7ff11152d6594d33b9c958a6a4b57872$$anon$2$$anon$1  - (Session.scala:113)
```

This uses reflection to compile the code at runtime, and slows down the initialization process.
To avoid this, we should generate a code with Scala macros in advance when binding a singleton in the above example.

**Comment by xerial** [comment]
Added a code to check runtime compilation time and it took around 0.5 ~ 1.0 sec: 
{code}
Compiling a code for embedding Session to JSONWebLogGenerator:
new (wvlet.airframe.Session => Any) {
  def apply(session:wvlet.airframe.Session) = {
    new com.treasuredata.prestobase.td.JSONWebLogGenerator {
      protected def __current_session = session
    }
  }
}    - (SessionImpl.scala:201)
2016-09-09 11:56:08-0700 trace [wvlet.airframe.SessionImpl] Compilation done: 0.64 sec.  - (SessionImpl.scala:207)
{code}

**Comment by xerial** [comment]
This occurs when building singletons when initializing a session. At this time, there is no TypeTag information so Design needs to prepare the above code in advance. 

We may need to extract the macro code to another project since macro cannot be used within the same library.

**Comment by xerial** [comment]
Fixed in #27. The above session injection code will be generated when you call design.bind[X] or subsequent toEagerSingleton[Y], etc.
